### PR TITLE
[hive] Partition is dropped from metastore only when all branches do not contain this partition

### PR DESCRIPTION
**Pull request by tsreaper**
### Purpose

Currently, when a partition is dropped from any branch, it will also be deleted from the metastore. This is inconvenient for the users because this partition may exist in other branches, including the main branch.

This PR fixes this issue. Paritition is now dropped from metastore only when all branches do not contain this partition.

### Tests

* Unit tests.

### API and Format

No changes in format.

### Documentation

No new feature.

### CODE REVIEW

File: paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/HiveCatalog.java [hunk 1]
```diff
@@ -304,15 +323,38 @@ && new CoreOptions(tableSchema.options()).partitionedTableInMetastore()) {
...
+    private boolean partitionExistsInOtherBranches(
+            Identifier identifier, Map<String, String> partitionSpec)
+            throws TableNotExistException {
+        FileStoreTable mainTable =
+                (FileStoreTable)
+                        getTable(
+                                new Identifier(
+                                        identifier.getDatabaseName(), identifier.getTableName()));
+        List<String> branchNames =
+                mainTable.branchManager().branches().stream()
```
Comment (JingsongLi) [inline paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/HiveCatalog.java:335]: It seems that we can optimize `branches` method, it is really costly.

File: paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/HiveCatalog.java [hunk 2]
```diff
@@ -304,15 +323,38 @@ && new CoreOptions(tableSchema.options()).partitionedTableInMetastore()) {
...
+                    (FileStoreTable)
+                            getTable(
+                                    new Identifier(
+                                            identifier.getDatabaseName(),
+                                            identifier.getTableName(),
+                                            branchName,
+                                            null));
+            if (!table.newScan().withPartitionFilter(partitionSpec).plan().splits().isEmpty()) {
+                return true;
+            }
```
Comment (LinMingQiang) [inline paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/HiveCatalog.java:355]: Can we add a method to BranchManager to get the partition information?

File: paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/HiveCatalog.java [hunk 3]
```diff
@@ -304,15 +322,35 @@ && new CoreOptions(tableSchema.options()).partitionedTableInMetastore()) {
...
+
+            FileStoreTable table =
+                    (FileStoreTable)
+                            getTable(
+                                    new Identifier(
+                                            identifier.getDatabaseName(),
+                                            identifier.getTableName(),
+                                            branchName,
+                                            null));
+            if (!table.newScan().withPartitionFilter(partitionSpec).plan().splits().isEmpty()) {
```
Comment (JingsongLi) [inline paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/HiveCatalog.java:349]: Use `TableScan.listPartitions()`.

File: paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/HiveCatalog.java [hunk 4]
```diff
@@ -304,15 +322,35 @@ && new CoreOptions(tableSchema.options()).partitionedTableInMetastore()) {
...
+        branchNames.add(DEFAULT_MAIN_BRANCH);
+
+        for (String branchName : branchNames) {
+            if (branchName.equals(identifier.getBranchNameOrDefault())) {
+                continue;
+            }
+
+            FileStoreTable table =
+                    (FileStoreTable)
+                            getTable(
```
Comment (JingsongLi) [inline paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/HiveCatalog.java:343]: Construct branch table from original table to improve performance.

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**